File: ninjaEyeDetector/ninjaEyeDetector.py
import numpy as np
import cv2

# Load the Cascade Classifier Xml file
face_cascade = cv2.CascadeClassifier("cascade/mallick_haarcascade_frontalface_default.xml")

# Specifying minimum and maximum size parameters
MIN_FACE_SIZE = 100
MAX_FACE_SIZE = 300

#Create a VideoCapture object
cap = cv2.VideoCapture(0)
 
# Check if camera opened successfully
if (cap.isOpened() == False): 
  print("Unable to read camera feed")

while(True):
  # Reading each frame
  ret, frameBig = cap.read()
 
  # If frame opened successfully
  if ret == True: 
     
    # Fixing the scaling factor
    scale = 640.0/frameBig.shape[1]
    
    # Resizing the image
    frame = cv2.resize(frameBig,None,fx = scale,fy = scale, interpolation = cv2.INTER_LINEAR)
  
    # Converting to grayscale
    frameGray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)

    # Detect faces
    faces = face_cascade.detectMultiScale(frameGray, scaleFactor=1.1, minNeighbors=5,flags=0, minSize=(MIN_FACE_SIZE,MIN_FACE_SIZE),maxSize=(MAX_FACE_SIZE,MAX_FACE_SIZE))
  
    # Loop over each detected face
    for i in xrange (0,len(faces)):
    
      # Dimension parameters for bounding rectangle for face
      x,y,width,height = faces[i];

      # Calculating the dimension parameters for eyes from the dimensions parameters of the face
      ex,ey,ewidth,eheight = int(x + 0.125*width), int(y + 0.25 * height), int(0.75 * width), int(0.25 * height)             
      
      # Drawing the bounding rectangle around the face
      cv2.rectangle(frame, (ex,ey),(ex+ewidth,ey+eheight),(128,255,0), 2)    

    # Display the resulting frame    
    cv2.imshow('Ninja Eye Detector',frame)
 
    # Press ESC on keyboard to stop tracking
    key = cv2.waitKey(1)
 
    if (key==27):

      break
 
  # Break the loop
  else:
    break 
# Release VideoCapture object
cap.release()

# Closes all the frames
cv2.destroyAllWindows() 

File: ninjaEyeDetector/ninjaEyeDetector.cpp
#include "opencv2/opencv.hpp" 

using namespace cv;
using namespace std;

// Specifying minimum and maximum size parameters
#define MIN_FACE_SIZE 100
#define MAX_FACE_SIZE 300

int main(void){
  // Load the Cascade Classifier Xml file
  CascadeClassifier faceCascade("cascade/mallick_haarcascade_frontalface_default.xml");
  
  // Create a VideoCapture object
  VideoCapture cap;

  // Check if camera opened successfully
  if(!cap.open(0)) return 0;
  
  Mat frame, frameBig, frameGray;
   
  while (1){
    
    // Reading each frame
    bool frameRead = cap.read(frameBig);

    // If frame not opened successfully
    if (!frameRead)
      break;

    // Fixing the scaling factor
    float scale = 640.0f/frameBig.cols;

    // Resizing the image
    resize(frameBig, frame, Size(), scale, scale);

    // Converting to grayscale
    cvtColor(frame, frameGray, COLOR_BGR2GRAY);

    // Creating vector to store the detected faces' parameters
    vector<Rect> faces;

    // Detect faces
    faceCascade.detectMultiScale(frameGray,faces,1.1,5,0,Size(MIN_FACE_SIZE,MIN_FACE_SIZE),Size(MAX_FACE_SIZE,MAX_FACE_SIZE));
    
    // Loop over each detected face
    for ( int i = 0; i < faces.size(); i++)
    {
      // Dimension parameters for bounding rectangle for face
      Rect faceRect = faces[i];

      // Calculating the dimension parameters for eyes from the dimensions parameters of the face
      Rect eyesRect = Rect(faceRect.x + 0.125*faceRect.width,faceRect.y + 0.25 * faceRect.height,0.75 * faceRect.width,
                       0.25 * faceRect.height);
      
      // Drawing the bounding rectangle around the face
      rectangle(frame, eyesRect,Scalar(128,255,0), 2);
      
    }
    
    // Display the resulting frame    
    imshow("Ninja Eye Detector", frame);
    int k = waitKey(1);

    // Press ESC on keyboard to stop tracking
    if(k == 27)
      break;
  }
  // release the VideoCapture object
  cap.release();
  
  // Closes all the windows
  destroyAllWindows();
  return 0;
}

